File: 3. Climbing Stairs.cpp
// 3. Climbing Stairs
// Topic: Climbing Stairs
// Description: Count number of distinct ways to reach step n using recursion + memoization.
// Input: n (integer >= 0)
// Output: number of ways

#include <bits/stdc++.h>
using namespace std;

// Memoization starts here: vector<int> memo
int ways(int n, vector<int>& memo)
{
    if (n == 0)
        return 1;
    if (n < 0)
        return 0;
    if (memo[n] != -1)
        return memo[n];
    memo[n] = ways(n - 1, memo) + ways(n - 2, memo);
    return memo[n];
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n))
        return 0;
    vector<int> memo(n + 1, -1);
    cout << ways(n, memo) << '\n';
}

// Memoization process started inside ways function where memo vector is used.
// Time Complexity: O(n)
// Space Complexity: O(n)

File: 3. Climbing Stairs.py
# 3. Climbing Stairs
# Topic: Climbing Stairs
# Description: Count number of distinct ways to reach step n using recursion + memoization.
# Input: n (integer >= 0)
# Output: number of ways

import sys
from functools import lru_cache

@lru_cache(maxsize=None)
def ways(n: int) -> int:
    if n == 0:
        return 1
    if n < 0:
        return 0
    return ways(n-1) + ways(n-2)

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    n = int(data[0])
    print(ways(n))

# Memoization process started at ways decorated with lru_cache.
# Time Complexity: O(n)
# Space Complexity: O(n)